packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,71 +89,71 @@ export class WorkspaceService implements FrontendApplicationContribution {
     }
 
     protected async updateWorkspace(): Promise<void> {
-        if (!this._workspaceFolder) {
-            const rootUri = await this.server.getMostRecentlyUsedWorkspace();
-            this._workspaceFolder = await this.toValidRoot(rootUri);
-            if (this._workspaceFolder) {
-                this._roots.push(this._workspaceFolder);
-            }
-        }
-        this.deferredRoots.resolve(this._roots);
+        await this.ensureWorkspaceInitialized();
+        await this.updateRoots();
+        this.watchRoots();
+    }
 
-        await this.preferences.ready;
+    protected async ensureWorkspaceInitialized(): Promise<void> {
         if (this._workspaceFolder) {
-            let roots: string[];
-            if (this.preferences['workspace.supportMultiRootWorkspace']) {
-                const rootConfig = await this.getRootConfig();
-                roots = rootConfig.roots;
-            } else {
-                roots = [this._workspaceFolder.uri];
-            }
+            return;
+        }
+        const rootUri = await this.server.getMostRecentlyUsedWorkspace();
+        this._workspaceFolder = await this.toValidRoot(rootUri);
+        const configUri = this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            this.watcher.watchFileChanges(configUri);
+        }
+    }
 
-            for (const rootBeingWatched of Object.keys(this.rootWatchers)) {
-                if (roots.indexOf(rootBeingWatched) < 0) {
-                    this.stopWatch(rootBeingWatched);
-                }
-            }
-            this._roots.length = 0;
-            for (const rootToWatch of roots) {
-                const valid = await this.toValidRoot(rootToWatch);
-                if (!this.rootWatchers[rootToWatch]) {
-                    await this.startWatch(valid);
-                }
-                if (valid) {
-                    this._roots.push(valid);
-                }
-            }
-            if (!this.rootWatchers[this._workspaceFolder.uri]) {
-                // must watch the workspace folder for meta data changes, even if it is not in the workspace
-                await this.startWatch(this._workspaceFolder);
+    protected async updateRoots(): Promise<void> {
+        this._roots = await this.computeRoots();
+        this.deferredRoots.resolve(this._roots); // in order to resolve first
+        this.deferredRoots = new Deferred<FileStat[]>();
+        this.deferredRoots.resolve(this._roots);
```

Comment:
could you please explain why we need to resolve it twice ? Thank you !

Reply:
The first resolve is only needed to resolve a promise created in the constructor, otherwise, it won't be ever resolved. I don't like it much either.

I don't see it much in other APIs. Usually, I use a promise if a single action should be awaited, but if there is a value which changes over the time, like `_roots` then synchronous access + an event is usually enough. We could analyze clients to see whether we can get rid of a  promise.